sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -138,56 +136,52 @@ object ConstantPropagation extends Rule[LogicalPlan] {
         if safeToReplace(right, nullIsFalse) =>
-        (None, Seq(((right, left), e)))
-      case a: And =>
-        val (newLeft, equalityPredicatesLeft) =
-          traverse(a.left, replaceChildren = false, nullIsFalse)
+        e -> mutable.Map(right.canonicalized -> (left, e))
+      case a @ And(left, right) =>
+        val (newLeft, equalityPredicates) =
+          traverse(left, replaceChildren = false, nullIsFalse)
         val (newRight, equalityPredicatesRight) =
-          traverse(a.right, replaceChildren = false, nullIsFalse)
-        val equalityPredicates = equalityPredicatesLeft ++ equalityPredicatesRight
-        val newSelf = if (equalityPredicates.nonEmpty && replaceChildren) {
-          Some(And(replaceConstants(newLeft.getOrElse(a.left), equalityPredicates),
-            replaceConstants(newRight.getOrElse(a.right), equalityPredicates)))
+          traverse(right, replaceChildren = false, nullIsFalse)
+        // We could recognize when conflicting constants are coming from the left and right sides
+        // and immediately shortcut the `And` expression to `Literal.FalseLiteral`, but that case is
+        // not so common and actually it is the job of `ConstantFolding` and `BooleanSimplification`
+        // rules to deal with those optimizations.
+        equalityPredicates ++= equalityPredicatesRight
+        val newAnd = a.withNewChildren(if (equalityPredicates.nonEmpty && replaceChildren) {
+          val replacedNewLeft = replaceConstants(newLeft, equalityPredicates)
+          val replacedNewRight = replaceConstants(newRight, equalityPredicates)
+          Seq(replacedNewLeft, replacedNewRight)
         } else {
-          if (newLeft.isDefined || newRight.isDefined) {
-            Some(And(newLeft.getOrElse(a.left), newRight.getOrElse(a.right)))
-          } else {
-            None
-          }
-        }
-        (newSelf, equalityPredicates)
+          Seq(newLeft, newRight)
+        })
+        newAnd -> equalityPredicates
       case o: Or =>
         // Ignore the EqualityPredicates from children since they are only propagated through And.
-        val (newLeft, _) = traverse(o.left, replaceChildren = true, nullIsFalse)
-        val (newRight, _) = traverse(o.right, replaceChildren = true, nullIsFalse)
-        val newSelf = if (newLeft.isDefined || newRight.isDefined) {
-          Some(Or(left = newLeft.getOrElse(o.left), right = newRight.getOrElse((o.right))))
-        } else {
-          None
-        }
-        (newSelf, Seq.empty)
+        o.mapChildren(traverse(_, replaceChildren = true, nullIsFalse)._1) -> mutable.Map.empty
       case n: Not =>
         // Ignore the EqualityPredicates from children since they are only propagated through And.
-        val (newChild, _) = traverse(n.child, replaceChildren = true, nullIsFalse = false)
-        (newChild.map(Not), Seq.empty)
-      case _ => (None, Seq.empty)
+        n.mapChildren(traverse(_, replaceChildren = true, nullIsFalse = false)._1) ->
```

Comment:
```scala
val newNot = n.mapChildren(traverse(_, replaceChildren = true, nullIsFalse = false)._1)
newNot -> AttributeMap.empty
```
?

Reply:
Ok, fixed in https://github.com/apache/spark/pull/40268/commits/13872471225799a4993d5544c50f68a3c2af2316.